common/cache/budget.go

```diff
@@ -0,0 +1,1260 @@
+		m.scope.UpdateGauge(metrics.BudgetManagerCapacityCount, float64(capacityCount))
+	}
+
+	// Emit usage metrics
+	m.scope.UpdateGauge(metrics.BudgetManagerUsedBytes, float64(m.UsedBytes()))
+	m.scope.UpdateGauge(metrics.BudgetManagerUsedCount, float64(m.UsedCount()))
+
+	// Emit soft threshold
+	m.scope.UpdateGauge(metrics.BudgetManagerSoftThreshold, m.enforcementSoftCapThreshold())
+
+	// Emit active cache count
+	m.scope.UpdateGauge(metrics.BudgetManagerActiveCacheCount, float64(atomic.LoadInt64(&m.activeCacheCount)))
+}
+
+// emitHardCapExceeded logs when hard capacity limit is exceeded and increments counter
+func (m *manager) emitHardCapExceeded(cacheID, budgetType string, requested, available uint64) {
+	if m.scope != nil {
+		m.scope.IncCounter(metrics.BudgetManagerHardCapExceeded)
+	}
+
+	if m.logger != nil {
+		m.logger.Debug("Hard capacity limit exceeded",
+			tag.Name(m.name),
+			tag.Dynamic("cache_id", cacheID),
+			tag.Value(budgetType),
+			tag.Counter(int(requested)),
+			tag.Number(int64(available)),
+		)
+	}
+}
+
+// emitSoftCapExceeded logs when soft cap is exceeded and increments counter
+func (m *manager) emitSoftCapExceeded(cacheID, budgetType string, requested, available uint64) {
+	if m.scope != nil {
+		m.scope.IncCounter(metrics.BudgetManagerSoftCapExceeded)
+	}
+
+	if m.logger != nil {
+		m.logger.Warn("Soft capacity limit exceeded",
+			tag.Name(m.name),
+			tag.Dynamic("cache_id", cacheID),
+			tag.Value(budgetType),
+			tag.Counter(int(requested)),
+			tag.Number(int64(available)),
+		)
+	}
+}
+
+// emitCapacityExceeded emits appropriate metrics based on the error type
+func (m *manager) emitCapacityExceeded(cacheID string, err error, requestedBytes uint64, requestedCount int64, capResult CapEnforcementResult) {
+	switch err {
+	case ErrBytesBudgetExceeded:
+		m.emitHardCapExceeded(cacheID, budgetTypeBytes, requestedBytes, capResult.AvailableBytes)
+	case ErrCountBudgetExceeded:
+		m.emitHardCapExceeded(cacheID, budgetTypeCount, uint64(requestedCount), uint64(capResult.AvailableCount))
+	case ErrBytesSoftCapExceeded:
+		m.emitSoftCapExceeded(cacheID, budgetTypeBytes, requestedBytes, capResult.AvailableBytes)
+	case ErrCountSoftCapExceeded:
+		m.emitSoftCapExceeded(cacheID, budgetTypeCount, uint64(requestedCount), uint64(capResult.AvailableCount))
+	}
+}
+
+// Cache-aware Reserve methods implementing two-tier soft cap logic
+
+func (m *manager) ReserveForCache(cacheID string, nBytes uint64, nCount int64) error {
+	// Check capacity constraints (both hard and soft cap)
+	capResult := m.enforceCapForCache(cacheID, nBytes, nCount)
+	if capResult.Error != nil {
+		m.emitCapacityExceeded(cacheID, capResult.Error, nBytes, nCount, capResult)
+		return capResult.Error
+	}
+
+	cacheUsage := m.getCacheUsage(cacheID)
+
+	cacheUsage.mu.Lock()
+	wasActive := cacheUsage.usedBytes > 0 || cacheUsage.usedCount > 0
+
+	if nBytes > 0 {
+		if err := m.reserveBytes(nBytes); err != nil {
+			cacheUsage.mu.Unlock()
```

Comment:
for future safety, do you mind moving this as a defer instead, so that it can't be missed on a refactor?